sled-agent/src/sled_agent.rs

```diff
@@ -4,46 +4,125 @@
-    external::Error, internal::nexus::DiskRuntimeState,
-    internal::nexus::InstanceRuntimeState,
+    internal::nexus::DiskRuntimeState, internal::nexus::InstanceRuntimeState,
     internal::sled_agent::InstanceHardware,
     internal::sled_agent::InstanceRuntimeStateRequested,
 };
-
 use slog::Logger;
 use std::sync::Arc;
 use uuid::Uuid;
 
-#[cfg(test)]
-use crate::mocks::MockNexusClient as NexusClient;
 #[cfg(not(test))]
-use nexus_client::Client as NexusClient;
+use {
+    crate::illumos::{dladm::Dladm, zfs::Zfs, zone::Zones},
+    nexus_client::Client as NexusClient,
+};
+#[cfg(test)]
+use {
+    crate::illumos::{
+        dladm::MockDladm as Dladm, zfs::MockZfs as Zfs,
+        zone::MockZones as Zones,
+    },
+    crate::mocks::MockNexusClient as NexusClient,
+};
 
-use crate::common::vlan::VlanID;
-use crate::instance_manager::InstanceManager;
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    Datalink(#[from] crate::illumos::dladm::Error),
+
+    #[error(transparent)]
+    Zone(#[from] crate::illumos::zone::Error),
+
+    #[error(transparent)]
+    Zfs(#[from] crate::illumos::zfs::Error),
+
+    #[error("Error managing instances: {0}")]
+    Instance(#[from] crate::instance_manager::Error),
+
+    #[error("Error managing storage: {0}")]
+    Storage(#[from] crate::storage_manager::Error),
+}
+
+impl From<Error> for omicron_common::api::external::Error {
+    fn from(err: Error) -> Self {
+        omicron_common::api::external::Error::InternalError {
+            internal_message: err.to_string(),
+        }
+    }
+}
 
 /// Describes an executing Sled Agent object.
 ///
 /// Contains both a connection to the Nexus, as well as managed instances.
 pub struct SledAgent {
+    _storage: StorageManager,
     instances: InstanceManager,
 }
 
 impl SledAgent {
     /// Initializes a new [`SledAgent`] object.
-    pub fn new(
-        id: &Uuid,
+    pub async fn new(
+        config: &Config,
         log: Logger,
-        vlan: Option<VlanID>,
         nexus_client: Arc<NexusClient>,
     ) -> Result<SledAgent, Error> {
+        let id = &config.id;
+        let vlan = config.vlan;
         info!(&log, "created sled agent"; "id" => ?id);
 
-        let instances = InstanceManager::new(log, vlan, nexus_client)?;
+        // Before we start creating zones, we need to ensure that the
+        // necessary ZFS and Zone resources are ready.
+        Zfs::ensure_filesystem(
+            ZONE_ZFS_DATASET,
+            Mountpoint::Path(std::path::PathBuf::from(
+                ZONE_ZFS_DATASET_MOUNTPOINT,
+            )),
+        )?;
+
+        // Identify all existing zones which should be managed by the Sled
+        // Agent.
+        //
+        // NOTE: Currently, we're removing these zones. In the future, we should
+        // re-establish contact (i.e., if the Sled Agent crashed, but we wanted
+        // to leave the running Zones intact).
+        let zones = Zones::get_non_base_zones()?;
```

Comment:
This is the only minor tweak I made post-approval. Basically, I'm only deleting the "non-base zones" at the moment.

I noticed a failure case locally where "deleting the base zone before an instance zone" results in a ZFS error that the filesystem has dependent clones. I think it's important to delete "non-base" zones before "base zones", at least for now without re-organizing the parent relationships.

All this is a bit moot in the long-term anyway, since we *shouldn't* be deleting these on reboot, but I'm just doing it for developer convenience.